### PR TITLE
Fixes VSTS Bug 1002614: System.ObjectDisposedException exception in

### PR DESCRIPTION
**Pull request by mkrueger**
MonoDevelop.VersionControl.Git.GitRepository.EnsureInitialized()

https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1002614

Some operations need to be canceled on Dispose.

### CODE REVIEW

**Comment by mkrueger** [comment]
@monojenkins backport release-8.3
@monojenkins backport release-8.4

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs [hunk 1]
```diff
@@ -135,8 +135,11 @@ protected void ShutdownScheduler ()
 			}
 		}
 
+		CancellationTokenSource disposeTokenSource = new CancellationTokenSource ();
+
 		public void Dispose ()
 		{
+			disposeTokenSource.Cancel ();
```
Comment (Therzok) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs]: `disposeTokenSource` needs to be disposed too.

Comment (mkrueger) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs]: fixed.

**Comment by sevoku** [comment]
@monojenkins rebase

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs [hunk 2]
```diff
@@ -135,8 +135,12 @@ protected void ShutdownScheduler ()
...
```
Comment (sevoku) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs]: this needs a IsDisposed check, see the tests failing with `System.ObjectDisposedException : The CancellationTokenSource has been disposed` in https://devdiv.visualstudio.com/DevDiv/_releaseProgress?_a=release-environment-logs&releaseId=502689&environmentId=2563893

Comment (mkrueger) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs]: fixed.

**Comment by sevoku** [comment]
@monojenkins backport release-8.3
@monojenkins backport release-8.4

**Pull request merged**